Warn about forgetting attention mask functions (#45811)
huggingface/transformers: docs/source/en/attention_interface.md

Customize or create new attention functions by adding them to the attention registry with [`AttentionInterface.register`]. Models use these functions through the `attn_implementation` argument.

> [!WARNING]

> > If you register a new attention function, you likely also need to register a new attention _mask_ function. If you do not register a new attention _mask_ function, **the mask is set to `None`, i.e., completely ignored**. This could break your model in unexpected ways (e.g., by silently discarding Causal Masking).

Reviewer: i think this wording may be a bit more accurate?

```suggestion
> [!WARNING]
> Register a matching attention mask function when you register a custom attention function. If the custom `attn_implementation` name is not registered in [`AttentionMaskInterface`], Transformers skips mask creation and passes `attention_mask=None` to the attention layers. Your attention function must handle causal, padding, packing, or sliding-window constraints itself, or those constraints can be silently dropped.
```

Author: Definitely, thanks!

I have reworded the suggestion a bit, to make it more clear to me when which consequences are to be expected; however, feel free to change it back to your original suggestion or modify it further as you see fit.

> This example customizes the attention function to print a statement for each layer (and keeps whatever mask is used in the original implementation, by registering [`masking_utils.sdpa_mask`] as attention mask function):

Reviewer: ```suggestion
This example customizes the attention function to print a statement for each layer. It keeps the mask in the original implementation by registering `masking_utils.sdpa_mask` as the attention mask function.
```
